test/Sema/diag_ambiguous_overloads.swift

```diff
@@ -19,19 +19,17 @@ func fg<T>(_ f: (T) -> T) -> Void {} // expected-note {{in call to function 'fg'
 fg({x in x}) // expected-error {{generic parameter 'T' could not be inferred}}
 
 
-// FIXME: Both f & g should complain about ambiguity of arg1, but in both cases the generic member
-// never gets into the list of candidates in the first place.
 struct S {
```

Comment:
You can look into why we aren't picking up these callees by examining the overload set passed to the constructor for CalleeCandidateInfo.

Reply:
This is because the "fnExpr" passed to the constructor of CalleeCandidateInfo is redefined in [this line](https://github.com/apple/swift/blob/master/lib/Sema/CSDiag.cpp#L5626), where the function subexpression is handled INDEPENDENTLY (i.e., without considering the actual argument information). In this case, the CS resolves an easy form of the function (e.g., the less formal parameters it has, the greater it is ranked). It may not be the best choice for diagnostics.

The motivation if this PR is to give it a chance to decide if we want to use the redefined "fnExpr" or the original one, taking into account the number of arguments.

After applying this PR, such FIXME would have been fixed, with the compiler raising better diagnostics.